Add GitHub actions workflow to test-build pull request branches

bjornoleh opened the pull request:
The new `.github/workflows/PR_test_build.yml` workflow checks out the pull request number that is entered as a repository variable.

It is possible to add an optional bundle ID suffix to not overwrite the TestFlight “production app” that one might be using live.

The checkout action starts with nightscout/Trio dev when setting up the git repository, so that the workflow is independent of any changes in the private fork.

## Setup:
Add repository variables:

### Mandatory:
- PR_NUMBER (Pull-request number in the nightscout/Trio repository)

### Optional:
- BUNDLE_ID_SUFFIX (appended to BUNDLE_IDENTIFIER = org.nightscout.$(DEVELOPMENT_TEAM).trio, which must match the bundle ID which is associated with the TestFlight app you wish to build to)

## Usage: 
- Run the workflow named “Test-build pull request”
- Wait for the build to complete and be processed at Apple
-  Install/update Trio in the TestFlight app. 
- The app display name will include the PR number (Trio_PRXXX) to easily distinguish it from other instances if the app one might have installed, or to see that the production app is in fact a PR build

Resolves #315 


---

**PR summary**

```
Add workflow to build and deploy pull request to TestFlight

This new GitHub Actions workflow (`PR_test_build.yml`) automates the process of building pull requests from the `nightscout/Trio` repository and deploying them to TestFlight.

Key features and changes:
- Adds a new workflow triggered manually via `workflow_dispatch`.
- Requires repository variable `PR_NUMBER` to specify the pull request to build.
- Optional variable `BUNDLE_ID_SUFFIX` can be used to modify the bundle identifier for test builds.
- In-place patching of BUNDLE_IDENTIFIER in `Config.xcconfig`
- Injects `APP_DISPLAY_NAME = Trio_PR<PR_NUMBER>` automatically to identify the app name as a PR build.
- If `BUNDLE_ID_SUFFIX` is set, updates `BUNDLE_IDENTIFIER` in place accordingly.
- Includes debug output for final `APP_DISPLAY_NAME` and `BUNDLE_IDENTIFIER`.

This enables easier, isolated testing of PRs directly via TestFlight with clearly labeled builds.
```

bjornoleh committed:
Create `PR_test_build.yml` to test-build pull requests with TestFlight

Usage:
Add repository variables:

Mandatory:
- PR_NUMBER (Pull-request number in the nightscout/Trio repository)

Optional:
- BUNDLE_ID_SUFFIX (appended to the default `org.nightscout.$(DEVELOPMENT_TEAM).trio` bundle ID, to be consistent with the bundle ID of the TestFlight app you wish to build to)
PR_test_build.yml: Start with checkout of nightscout/Trio dev

The PR number is appended to APP_DISPLAY_NAME for easy identification of PR builds on the phone.

dnzxy commented:
Caveat: I personally don't think users that cannot git should be running PR builds and test unreleased features.
This action does appear to do what it is intended to, but it is behind the April 24, 2025 deadline, i.e. wrong Xcode and MacOS runner version. Feel free to update the PR accordingly, I'm still not much in favor of it. 
Depending on how others feel about it, we can definitely re-visit it though :)

bjornoleh commented:
I’ll synccthe branch with dev once TCD is here, then we can decide. 

It’s not so much about catering for people who can’t git, but a very convenient way of testing that GH actions works for various PRs, and also to test PRs without having a MAC available at the moment.

t1dude commented:
This would be a nice feature for a developer or an advanced tester, but I don't think this should be a feature in any main release of Trio, as it could encourage users to build several app instances with unfinished PR's. Is there a way to implement this without making it available to every "normal" user? Separate branch? Customization script?

bjornoleh commented:
I just synced this with latest dev. There's been some negative feedback on this due to potential risk of misuse by regular users. Maybe we could make this require some "secret handshake" that only the devs know? 😄

MikePlante1 commented:
> I just synced this with latest dev. There's been some negative feedback on this due to potential risk of misuse by regular users. Maybe we could make this require some "secret handshake" that only the devs know? 😄

I just don’t see a purpose. It isn’t difficult to just make a PR to your own repo from the branch/repo the PR is being made from.

bjornoleh committed:
Optionally revert commits if vars.REVERT_COMMITS is set

Usage: Add the repository variable `REVERT_COMMITS` and add commit IDs to be reverted.

Multiple commits can be reverted. Add one commit per line, with optional trailing #comments for identification of the commit hashes.

Commented out lines are ignored.

Changed the checkout repo action to use `fetch-depth: 0` to allow fetching the full commit history (required for reverting commits).

bjornoleh commented:
I rebased the PR onto current dev, and squashed some commits.

I also added an optional `REVERT_COMMITS` repository variable, that became useful for my own testing now due to a commit that broke my particular use case for Dexcom Share backfills with G6. Now I can easily comment out that change by setting a repo variable, instead of actually commit changes to the various branches I might be testing.

```
Usage: Add the repository variable `REVERT_COMMITS` and add commit IDs to be reverted.

Multiple commits can be reverted. Add one commit per line, with optional trailing #comments for identification of the commit hashes.

Commented out lines are ignored.

Changed the checkout repo action to use `fetch-depth: 0` to allow fetching the full commit history (required for reverting commits).
```

dnzxy commented:
I still do not think this is a good idea to be in the "main" repository, sorry. 
We could add a how-to to TrioDocs with a template file and explanation on how to create this workflow file and to set it up.
I'd suggest to close this here, and if still wanted / deemed useful, move this to docs, as suggested.

bjornoleh commented:
All good, will close the PR but keep the branch in case someone stumbles upon this at a later time.
